core/executor/src/precompiles/modexp.rs

```diff
@@ -64,17 +63,34 @@ impl PrecompileContract for ModExp {
     }
 
     fn gas_cost(input: &[u8]) -> u64 {
-        match LargeNumber::parse(input) {
-            Ok(large_number) => {
-                let dynamic_gas =
-                    large_number.multiplication_complexity() * large_number.iterator_count() / 3u64;
-
-                dynamic_gas
-                    .max(Integer::from(Self::MIN_GAS))
-                    .saturating_as()
-            }
-            Err(_) => u64::MAX,
-        }
+        let calc = || -> Result<u64, PrecompileFailure> {
+            let base_size = get_data(input, 0, 32)?.saturating_as::<usize>();
+            let exponent_size = get_data(input, 32, 32)?.saturating_as::<usize>();
+            let modulo_size = get_data(input, 64, 32)?.saturating_as::<usize>();
+
+            let data = if input.len() > 96 {
+                &input[96..]
+            } else {
+                &input[0..0]
+            };
+
+            let exponent = if exponent_size >= 32 {
+                get_data(data, base_size, 32)?
+            } else {
+                get_data(data, base_size, exponent_size)?
+            };
+
+            let multiplication_complexity = multiplication_complexity(base_size, modulo_size);
+
+            let iterator_count = iterator_count(exponent_size, exponent);
+
+            let dynamic_gas = multiplication_complexity * iterator_count / 3u64;
+            Ok(dynamic_gas
+                .max(Integer::from(Self::MIN_GAS))
+                .saturating_as::<u64>())
+        };
+
+        calc().unwrap_or(u64::MAX)
```

Comment:
We shouldn't use `unwrap_or` here because `get_data` failing will be accidental. See also https://github.com/axonweb3/axon/issues/1287

Reply:
Similarly, we shouldn't convert allocation error to `ExitError::StackOverflow` because that'll also be non-determinism.

Reply:
We should just abort if allocation fails now. Big allocation is prevented by gas cost now.

Reply:
If any failure here, the return is to consume all the gas, and the interface definition cannot return an error. If you want to return an error, please choose to open another discussion, which is not within the scope of work of this PR

Reply:
Just don't handle allocation failure.